Official TypeScript Support RFC: draft for strike team feedback (#2)
chriskrycho/ember-rfcs: text/0000-road-to-typescript.md

Release Versions: Leave as is
Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
RFC PR: Fill this in with the URL for the Proposal RFC PR
-->

# Road to TypeScript

## Summary

TypeScript has become a key part of the front-end development ecosystem over the past several years, and powers many of the best developer experiences in the front-end ecosystem. Ember was a relatively early TypeScript adopter for its internals, and there is widespread usage in the ecosystem with community support, but to date Ember has not provided “out of the box” or official support for TypeScript. This RFC declares our intent to make TypeScript a first-class citizen of the Ember ecosystem, as a peer to JavaScript which makes the developer experience better for *all* Ember developers. It outlines the key constraints and goals for the effort, and outlines a roadmap for accomplishing those goals.

## Motivation

> Why are we doing this? What use cases does it support? What is the expected
outcome?

## Detailed design

TODO: give this an intro

Accordingly, the path to full TypeScript support in Ember requires RFCs to address each of the following points:

> - **What “official support” entails.** Does it require that all officially published libraries ship TypeScript types? Does it require a specific support policy for all officially published libraries? Should the RFC template be updated in any way to account for TypeScript API constraints? etc.

Author: Alternative approach: *this* RFC could define this. 🤔

- **Supported compilation modes.** What should the out-of-the-box settings be for compiling Ember apps and addons written in TypeScript? This will need to be written with an eye to Embroider and the v2 Package Format as end goals, but without assuming them as available at the starting point. It must also address the transition path from current ember-cli-typescript (if any). It *must* address interoperation hazards between Babel and TypeScript (for example, in the decorators implementations).

- **The `@glimmer/component@2.0.0` Type API.** The current Glimmer component TypeScript API works well, but research efforts into “typed templates” and [component documentation](https://github.com/emberjs/rfcs/pull/678) have both exposed the importance of being able to specify more “type parameters” for components than simply its args, including its root element(s) and blocks, if any. Additionally, the types should be readily extensible if other such concerns emerge in the future.

> - **Official Documentation.** It is important, if Ember officially supports TypeScript, that we support it in all of Ember’s documentation, including guides, API docs, etc. This requires considerable design effort, including 

Author: ```suggestion
- **Official Documentation.** It is important, if Ember officially supports TypeScript, that we support it in all of Ember’s documentation, including guides, API docs, etc. This requires considerable design effort, including thinking through how to present then two (side by side? toggleable? etc.), where to introduce discussion of TypeScript support in the guides, etc.
```

TypeScript explicitly *does not* follow SemVer. TypeScript's core team argues that *every* change to the compiler is a breaking change, and that SemVer is therefore meaningless. (We do not agree with this characterization, but are also uninterested in arguing it. This RFC takes the TypeScript team's position as a given.) Accordingly, every TypeScript point release may be a breaking change, and "major" numbers for releases signify nothing beyond having reached `x.9` in the previous cycle.

For TypeScript to be a first-class citizen of the Ember ecosystem, we need:

-   a policy defining what constitutes a breaking change for consumers of a library which publishes types, including Ember’s core libraries
-   tooling to detect breaking changes in types—whether from refactors, or from new TypeScript releases—and to minimize the amount of churn from breaking changes in TypeScript
-   a general and widely-adopted policy for supported TypeScript versions

Once all three of those elements are adopted, end users will be able to have equally high confidence in the stability of published types as they do in their runtime code.

### Non-Goals
#### Superseding or replacing JavaScript

> This RFC aims to *add* a new first-class language to the Ember support matrix. However, it does *not* recommend *replacing* JavaScript with TypeScript. To the contrary: this is an explicit non-goal, and a key design constraint for this RFC (as well as a long-standing commitment of the Typed Ember team) is that TypeScript support should benefit *everyone* who uses Ember, whether they ever write a single line of JavaScript.

Author: ```suggestion
This RFC aims to *add* a new first-class language to the Ember support matrix. However, it does *not* recommend replacing JavaScript with TypeScript. To the contrary: this is an explicit non-goal, and two key design constraints for this RFC (as well as long-standing commitments of the Typed Ember team) are:

- TypeScript support should benefit *everyone* who uses Ember, whether they ever write a single line of TypeScript
- Using TypeScript should never be mandatory for anyone who uses Ember
```

### Roadmap

With these constraints in mind, the path to full TypeScript support in Ember requires RFCs to address each of the following points:

- **What “official support” entails.** Does it require that all officially published libraries ship TypeScript types? Does it require a specific support policy for all officially published libraries? Should the RFC template be updated in any way to account for TypeScript API constraints? etc.

- **The set of supported types.** Specifically, should the types support the whole API surface of Ember including both Classic and Octane features, or only Octane features? For example: mixins are not part of the Octane programming model *and* are a major source of the complexity of Ember’s current types; should they be included at all in types shipped with 

- **Migration from DefinitelyTyped.** TypeScript users in the Ember community currently rely on the types in DefinitelyTyped. We need to define a transition story for them, which is closely related to the previous point. If we drop support for features currently supported on DefinitelyTyped, or make even well-motivated changes, how will users migrate successfully to them?

> - **Semantic Versioning of types and TypeScript compiler version support.** There is currently no widely-used definition for semantic versioning of TypeScript types. Additionally, we must define a support policy for TypeScript versions. (See also further discussion [below](#semantic-versioning-and-supported-typescript-versions).)

Author: Worth expanding the sentence on TypeScript versions to at least allude to the issues: 

```suggestion
- **Semantic Versioning of types and TypeScript compiler version support.** There is currently no widely-used definition for semantic versioning of TypeScript types; we must define the policy Ember uses—hopefully in a way which can benefit the broader TypeScript community. Additionally, we must define a support policy for TypeScript versions: do we pin to a minimum type when we release a major, do something our Node support policy, or some other approach? (See also further discussion [below](#semantic-versioning-and-supported-typescript-versions).)
```

> - **Supported compilation modes.** What should the out-of-the-box settings be for compiling Ember apps and addons written in TypeScript? This will need to be written with an eye to Embroider and the v2 Package Format as end goals, but without assuming them as available at the starting point. It must also address the transition path from current ember-cli-typescript (if any). It *must* address interoperation hazards between Babel and TypeScript (for example, in the decorators implementations).

Author: This should also call out `isolatedModules` as an obvious thing we need to address.

- **Official Documentation.** It is important, if Ember officially supports TypeScript, that we support it in all of Ember’s documentation, including guides, API docs, etc. This requires considerable design effort, including thinking through how to present them (side by side? toggle-able? etc.), where to introduce discussion of TypeScript support in the guides, etc.

There are also concerns which must be addressed, but do not require RFCs (they are key implementation details, but *are* implementation details):

- **How are the published types generated?** While much of Ember is implemented in TypeScript, its internal types are not currently ready to be used for public API, and it will likely take some time for them to be ready (with different packages ready at different times).

> - **How does the build tooling publish the types?** TypeScript generally assumes that types are published with packages of the same name. While there is [now](https://emberjs.github.io/rfcs/0706-deprecate-ember-global.html) a path to publishing actual packages like `"@ember/object"`, this effort should not be gated on any such effort, and there is additional work to be done. However

Author: This is an incomplete sentence. Needs to be a complete sentence. 😂

- **How do consumers *consume* published types?** There are a few “gotchas” about how TypeScript supports features like autocomplete in a project, which primarily affect the very first few interactions end users have with TypeScript, which we should address via tooling.

> 

Author: We may also need an RFC for adding something like a `--typescript`/`--ts` flag to `ember new`.

---
Stage: Accepted
Start Date: 2021-02-19
Release Date: Unreleased
Release Versions:
  ember-source: vX.Y.Z
  ember-data: vX.Y.Z
Relevant Team(s): Ember.js, Ember Data, Ember CLI, Learning, Steering
RFC PR: 
---

<!--- 
Directions for above: 

Stage: Leave as is
Start Date: Fill in with today's date, YYYY-MM-DD
Release Date: Leave as is
Release Versions: Leave as is
Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
RFC PR: Fill this in with the URL for the Proposal RFC PR
-->

# Road to TypeScript

## Summary

> TypeScript has become a key part of the front-end development ecosystem over the past several years, and powers many of the best developer experiences in the front-end ecosystem. Ember was a relatively early TypeScript adopter for its internals, and there is widespread usage in the ecosystem with community support, but to date Ember has not provided “out of the box” or official support for TypeScript. This RFC declares our intent to make TypeScript a first-class citizen of the Ember ecosystem, as a peer to JavaScript which makes the developer experience better for *all* Ember developers. It outlines the key constraints and goals for the effort, and outlines a roadmap for accomplishing those goals.

Reviewer: >  but to date Ember has not provided “out of the box” or official support for TypeScript

might be good to clarify that this is about "for app / addon developers"? not sure that is or isn't a good distinction but wanted to point it out in case you do want to tweak

Author: Resolved in 28a1fc4 – I like clarifying this!

## Motivation

> TypeScript is a key ingredient of contemporary front-end development. Per community surveys like the GitHub Octoverse, the Stack Overflow Developer report, and the State of JS survey, TypeScript has exploded in popularity over the last half decade, and is the rare tool which has continuously grown in both both usage and satisfaction over that time. ( While all of these surveys should be understood to be wildly unrepresentative of the broader front-end development community in various ways, they nonetheless provide some useful signal about the state of the ecosystem.)

Reviewer: Might be nice for the survey mentions to be links

Author: Added survey links and more industry info in a28dfb8.

Additionally, the development of the language server protocol and the widespread adoption of the TypeScript language server has led to first-class support for TypeScript across many tools, from Vim to Visual Studio. This support also powers rich developer experiences for JavaScript-only developers, as it can provide smart autocomplete, inline documentation, and more even for codebases which are written solely in JavaScript. Thus, non-TypeScript and TypeScript users alike benefit when working with a project which supplies type definitions.

Ember users have been working with TypeScript for years with community support led by the Typed Ember team, but there is one critical problem which cannot be solved without making TypeScript support official, and several key ways in which official support would substantially improve the experience of both TypeScript and JavaScript developers:

> 

Author: - [x] Make extra extra explicit the *benefits* which **everyone** gains.
- [x] Making TS part of the "safety of the herd" (I prefer different language, but the idea is good)

Author: Added further notes here in 3843459.

- [ ] the SemVer problem
    - DefinitelyTyped
    - TS version support policy
- [ ] other key benefits for official adoption
    - coordination (e.g. editions aren't done till docs also goes for TS and types)
    - [ ] blueprints?

>     - [ ] more?

Reviewer 2: a more immersive experience for TS user from the ember (cli) tools - such as the generators

Author: Integrated in 3843459.

## Detailed design

The primary goal of this RFC is to add TypeScript as an officially supported language. This involves four major elements:

- Identifying the [constraints](#constraints) under which adoption can proceed
- Making explicit the [non-goals](#non-goals) of this project
- [Defining official support](#official-support): what it does and does not entail
- Building a [roadmap](#roadmap) for completing the effort

### Constraints

The following are our hard constraints in supporting TypeScript with Ember:

- We must not compromise our commitment to “stability without stagnation” and our strong Semantic Versioning guarantees. 
- Using TypeScript should never be mandatory for anyone who uses Ember.

> - TypeScript support may never *degrade* the experience of JavaScript users, and wherever possible it should benefit both JavaScript and TypeScript developers.

Reviewer: > TypeScript support may never *degrade* the experience of JavaScript users

The "may" here reads quite oddly to me. I'd expect **must**...

Author: ```suggestion
- TypeScript support must never *degrade* the experience of JavaScript users, and wherever possible it should benefit both JavaScript and TypeScript developers.
```

> 

Author: Add: types are published *in their packages*.

Author: Resolved via 262722

> ### Non-Goals

Author: - [x] improve the definition of "non-goals": not included in initial road to TS
- [x] include "not even changing the default" in second section

Author: 54ac11d

The *only* relationship those efforts have to this roadmap is that they will inform the design of the `@glimmer/component@2.x` TypeScript API (as described below under **Roadmap**).

### Official Support

<!--

**What “official support” entails.** Does it require that all officially published libraries ship TypeScript types? Does it require a specific support policy for all officially published libraries? Should the RFC template be updated in any way to account for TypeScript API constraints? etc.

> -->

Author: - Libraries which are direct dependencies of the default blueprint and have public JavaScript APIs should ship their own types.
- Things like ember-template-lint (not an official library) may *not* need it:
     - not official
     - not a public JS API
- Ember CLI: intend to meet the same commitment, but not required for MVP b/c gated on v2 package format being the default including a well-defined authoring format

Reviewer 2: What are "officially published libraries" ?

Another idea from storybook team here: They grouped some of their addons as "essentials". Maybe adopting some strategy might help, too?

Author: I’ve clarified this in the latest revision. Can discuss further in the strike team meeting tomorrow!

- **Semantic Versioning of types and TypeScript compiler version support.** There is currently no widely-used definition for semantic versioning of TypeScript types; we must define the policy Ember uses—hopefully in a way which can benefit the broader TypeScript community. Additionally, we must define a support policy for TypeScript versions: do we pin to a minimum type when we release a major, do something our Node support policy, or some other approach? (See also further discussion [below](#semantic-versioning-and-supported-typescript-versions).)

- **Supported compilation modes.** What should the out-of-the-box settings be for compiling Ember apps and addons written in TypeScript? This will need to be written with an eye to Embroider and the v2 Package Format as end goals, but without assuming them as available at the starting point. It must also address the transition path from current ember-cli-typescript (if any). It *must* address interoperation hazards between Babel and TypeScript: for example, in the decorators implementations, use of the `isolateModules` setting and associated lints, etc.

> - **The `@glimmer/component@2.0.0` Type API.** The current Glimmer component TypeScript API works well, but research efforts into “typed templates” and [component documentation](https://github.com/emberjs/rfcs/pull/678) have both exposed the importance of being able to specify more “type parameters” for components than simply its args, including its root element(s) and blocks, if any. Additionally, the types should be readily extensible if other such concerns emerge in the future.

Reviewer: We should remove `isDestroying`/`isDestroyed`/`willDestroy`

Author: Called out the destroyables bit in 19ffd9f, with a clarification that it's not a blocker for TS adoption but likely makes sense to address at the same time.

> - **Official Documentation.** It is important, if Ember officially supports TypeScript, that we support it in all of Ember’s documentation, including guides, API docs, etc. This requires considerable design effort, including thinking through how to present them (side by side? toggle-able? etc.), where to introduce discussion of TypeScript support in the guides, etc.

Author: Update RFC template as well to mention TS

Author: Added RFC template comments in d5f9b7f

> - **Official Documentation.** It is important, if Ember officially supports TypeScript, that we support it in all of Ember’s documentation, including guides, API docs, etc. This requires considerable design effort, including thinking through how to present them (side by side? toggle-able? etc.), where to introduce discussion of TypeScript support in the guides, etc.

Reviewer 2: tabs for code snippets? With js and ts code?

Author: To be decided *in the docs RFC*. I'm intentionally leaving aside all of those details here. 😄

> - **How do consumers generate TypeScript apps and addons (and blueprints more generally)?** Today, users install `ember-cli-typescript` to opt into using TypeScript, and there is a wholly separate (and badly-maintained!) set of blueprints for working with TypeScript. Should we support `ember new --typescript`? How should blueprints work?

Author: Ember CLI, Embroider, v2 Package Format (both authoring and publishing), migration path from current `ember-cli-typescript`, etc.

Also note that our authoring policy will end up being inferred into other packages; we *can* (and maybe should) provide official *recommendations* about this.

Author: Resolved in 36ea5e4 and badeda1

There are also concerns which must be addressed, but do not require RFCs (they are key implementation details, but *are* implementation details):

> - **How are the published types generated?** While much of Ember is implemented in TypeScript, its internal types are not currently ready to be used for public API, and it will likely take some time for them to be ready (with different packages ready at different times).

Reviewer 2: Is this about `@ember/*` packages only? Or generic section about all packages, such as community packages?

Author: This is specific to the official types (as we will define them in the main discussion of "official libraries"; likely it'll be "all dependencies installed by the default blueprint" or something like that). We will make *recommendations* for the ecosystem, of course, but ultimately addon authors can (and will!) do as they like.

> - **How does the build tooling publish the types?** TypeScript generally assumes that types are published with packages of the same name. While there is [now](https://emberjs.github.io/rfcs/0706-deprecate-ember-global.html) a path to publishing actual packages like `"@ember/object"`, this effort should not be gated on any such effort, and there is additional work to be done. However, that work simply consists of implementation details, rather than defining new public API.

Reviewer 2: Same question here: Is this about `@ember/*` packages only? Or generic section about all packages, such as community packages?

Author: Same answer as above. 😉

> - **How do consumers *consume* published types?** There are a few “gotchas” about how TypeScript supports features like autocomplete in a project, which primarily affect the very first few interactions end users have with TypeScript, which we should address via tooling.

Reviewer 2: We made good experiences with `typesVersions` in `package.json` - which TS is picking up (atm I dunno if this is only within a monorepo or in general). Does this RFC need such a detail? Or is it enough to say "we provide tooling for this" - because this sentence is appearing quite often then.

If the latter, then it might be worth adding a section for the provided tooling support. Where will the home for that tooling be? What will it support developers with? and how to use it as a consumer

Author: This is also in the bucket of "to be decided in specific RFCs". This RFC will remain intentionally high-level, and individual RFCs will dig into the details as appropriate. Switching to something like the combination of composite projects and using `typesVersions` will very likely be among those details, but they don't belong in this roadmap-style RFC.

- tooling to detect breaking changes in types—whether from refactors, or from new TypeScript releases—and to minimize the amount of churn from breaking changes in TypeScript
- a general and widely-adopted policy for supported TypeScript versions

Once all three of those elements are adopted, end users will be able to have equally high confidence in the stability of published types as they do in their runtime code.

## How we teach this

This RFC intentionally defers most of the analysis of teaching work to an RFC dedicated to the question. At a high level, once the roadmap is complete, the Ember blog should announce support, and Ember’s docs should explicitly state that TypeScript is officially supported and show TypeScript examples alongside JavaScript examples.

It is critical that these materials (including the blog post) emphasize that TypeScript support is *additive*, not a replacement for JavaScript, as discussed above in [Non-Goals](#non-goals).

## Drawbacks

Adding TypeScript support imposes an additional maintenance burden on all contributors to Ember:

- documentation must be kept up to date in both JavaScript and TypeScript

> - changes to APIs must be kept in sync with published types

Reviewer 3: The above two points only apply to cases where the underlying Ember source has not been updated to TypeScript and we're maintaining separate `d.ts` files, correct?

Author: Certainly that is true for the second case. (I currently anticipate that will be a matter of years for a lot of the core types, though.) For documentation, it is a mix: you are correct as far as API docs go, but for guides, tutorials, etc., the concern will remain regardless.

Author: @wagenet I tweaked the wording here to clarify the difference between implementation-derived types and hand-authored types in 19ffd9f – thanks for calling that out!

- whether Ember should or should not aim to support TypeScript as a first-class citizen
- if it should adopt TypeScript, whether there are any gaps in the proposed roadmap

-->

### Outline <!-- omit in toc -->

- [Motivation](#motivation)
- [Detailed design](#detailed-design)
  - [Defining Official Support](#defining-official-support)
  - [Constraints](#constraints)
  - [Non-Goals](#non-goals)
    - [Superseding or replacing JavaScript](#superseding-or-replacing-javascript)
    - [Typed Templates](#typed-templates)
  - [Roadmap](#roadmap)
    - [RFC Required](#rfc-required)
      - [Types](#types)
      - [Official documentation](#official-documentation)
      - [Authoring and Build](#authoring-and-build)
    - [Implementation Details](#implementation-details)
    - [Recommendations](#recommendations)
  - [Semantic Versioning and Supported TypeScript Versions](#semantic-versioning-and-supported-typescript-versions)
- [How we teach this](#how-we-teach-this)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
- [Unresolved questions](#unresolved-questions)

## Motivation

> TypeScript is a key ingredient of contemporary front-end development. Per community surveys like [the GitHub Octoverse](https://octoverse.github.com), the [Stack Overflow Developer report](https://insights.stackoverflow.com/survey/2020), and the [State of JS survey](https://2020.stateofjs.com/en-US/), TypeScript has exploded in popularity over the last half decade, and is the rare tool which has continuously grown in both both usage and satisfaction over that time. While all of these surveys should be understood to be wildly unrepresentative of the broader front-end development community in various ways, they nonetheless provide some useful signal about the state of the ecosystem, especially when combined with experience reports from major companies like Microsoft (millions of lines of Typescript, all of it adopted voluntarily), Google, AirBnB, Dropbox, Bloomberg, Slack, and many others.

Reviewer 2: ```suggestion
TypeScript is a key ingredient of contemporary front-end development. Per community surveys like [the GitHub Octoverse](https://octoverse.github.com), the [Stack Overflow Developer report](https://insights.stackoverflow.com/survey/2020), and the [State of JS survey](https://2020.stateofjs.com/en-US/), TypeScript has exploded in popularity over the last half decade, and is the rare tool which has continuously grown in both usage and satisfaction over that time. While all of these surveys should be understood to be wildly unrepresentative of the broader front-end development community in various ways, they nonetheless provide some useful signal about the state of the ecosystem, especially when combined with experience reports from major companies like Microsoft (millions of lines of Typescript, all of it adopted voluntarily), Google, AirBnB, Dropbox, Bloomberg, Slack, and many others.
```

Author: Thank you!
